File: backend/app/crypto/keys.py
"""Ed25519 keys as base64url text: raw 32-byte seeds and public keys (docs/04 "Keys", D-014)."""

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey
from cryptography.hazmat.primitives.serialization import (
    Encoding,
    NoEncryption,
    PrivateFormat,
    PublicFormat,
)

from app.config import Settings
from app.crypto import base64url
from app.crypto.ed25519_point import is_prime_order_point

KEY_BYTES = 32


class InvalidKey(ValueError):
    """Key text is not base64url of a raw 32-byte Ed25519 key. Messages never include the text."""


class RootKeyNotConfigured(RuntimeError):
    """ROOT_SIGNING_KEY_B64 is not set."""


def _raw_key(text: str, what: str) -> bytes:
    try:
        raw = base64url.decode(text)
    except base64url.InvalidBase64:
        raise InvalidKey(f"{what} is not base64url without padding") from None
    if len(raw) != KEY_BYTES:
        raise InvalidKey(f"{what} must decode to {KEY_BYTES} bytes, got {len(raw)}")
    return raw


def private_key_from_seed(seed: str) -> Ed25519PrivateKey:
    return Ed25519PrivateKey.from_private_bytes(_raw_key(seed, "private key seed"))


def seed_b64url(key: Ed25519PrivateKey) -> str:
    return base64url.encode(key.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption()))


def public_key_from_b64url(text: str) -> Ed25519PublicKey:
    """Only canonical prime-order points: small-order keys allow keyless forgery (D-015)."""
    raw = _raw_key(text, "public key")
    if not is_prime_order_point(raw):
        raise InvalidKey("public key is not a canonical prime-order Ed25519 point")
    try:
        return Ed25519PublicKey.from_public_bytes(raw)
    except ValueError:
        raise InvalidKey("public key is not a valid Ed25519 key") from None


def public_key_b64url(key: Ed25519PublicKey) -> str:
    return base64url.encode(key.public_bytes(Encoding.Raw, PublicFormat.Raw))


def load_root_private_key(settings: Settings) -> Ed25519PrivateKey:
    """The root signing key from ROOT_SIGNING_KEY_B64. Never log it, never return it."""
    if settings.root_signing_key_b64 is None:
        raise RootKeyNotConfigured("ROOT_SIGNING_KEY_B64 is not set (see backend/.env.example)")
    return private_key_from_seed(settings.root_signing_key_b64.get_secret_value())

File: dashboard/src/lib/cert/keys.ts
import * as ed from "@noble/ed25519";
import * as b64 from "./base64url";

export class InvalidKey extends Error {}

/**
 * True only for the canonical encoding of an Ed25519 point of prime order L (D-015).
 * Small-order and mixed-order keys let anyone forge signatures that verify.
 */
export function isPrimeOrderPoint(raw: Uint8Array): boolean {
  if (raw.length !== 32) return false;
  try {
    const point = ed.Point.fromBytes(raw, false); // strict: rejects non-canonical encodings
    return !point.isSmallOrder() && point.isTorsionFree();
  } catch {
    return false;
  }
}

/** A public key from base64url, or InvalidKey. */
export function publicKeyFromB64url(text: string): Uint8Array {
  let raw: Uint8Array;
  try {
    raw = b64.decode(text);
  } catch {
    throw new InvalidKey("public key is not canonical base64url");
  }
  if (!isPrimeOrderPoint(raw)) throw new InvalidKey("not a prime-order Ed25519 point");
  return raw;
}

export function hexToBytes(hex: string): Uint8Array {
  return Uint8Array.from(hex.match(/../g) ?? [], (h) => parseInt(h, 16));
}
